libraries/Microsoft.Bot.Builder/Skills/SkillHandlerImpl.cs

```diff
@@ -0,0 +1,183 @@
+            turnContext.Activity.Entities = endOfConversationActivity.Entities;
+            turnContext.Activity.Locale = endOfConversationActivity.Locale;
+            turnContext.Activity.LocalTimestamp = endOfConversationActivity.LocalTimestamp;
+            turnContext.Activity.Timestamp = endOfConversationActivity.Timestamp;
+            turnContext.Activity.ChannelData = endOfConversationActivity.ChannelData;
+            turnContext.Activity.Properties = endOfConversationActivity.Properties;
+        }
+
+        private static void ApplyEventToTurnContextActivity(ITurnContext turnContext, Activity eventActivity)
+        {
+            // transform the turnContext.Activity to be the EventActivity.
+            turnContext.Activity.Type = eventActivity.Type;
+            turnContext.Activity.Name = eventActivity.Name;
+            turnContext.Activity.Value = eventActivity.Value;
+            turnContext.Activity.RelatesTo = eventActivity.RelatesTo;
+
+            turnContext.Activity.ReplyToId = eventActivity.ReplyToId;
+            turnContext.Activity.Value = eventActivity.Value;
+            turnContext.Activity.Entities = eventActivity.Entities;
+            turnContext.Activity.Locale = eventActivity.Locale;
+            turnContext.Activity.LocalTimestamp = eventActivity.LocalTimestamp;
+            turnContext.Activity.Timestamp = eventActivity.Timestamp;
+            turnContext.Activity.ChannelData = eventActivity.ChannelData;
+            turnContext.Activity.Properties = eventActivity.Properties;
+        }
+
+        private async Task<SkillConversationReference> GetSkillConversationReferenceAsync(string conversationId, CancellationToken cancellationToken)
+        {
+            SkillConversationReference skillConversationReference;
+            try
+            {
+                skillConversationReference = await _conversationIdFactory.GetSkillConversationReferenceAsync(conversationId, cancellationToken).ConfigureAwait(false);
+            }
+            catch (NotImplementedException)
+            {
+                _logger.LogWarning("Got NotImplementedException when trying to call GetSkillConversationReferenceAsync() on the ConversationIdFactory, attempting to use deprecated GetConversationReferenceAsync() method instead.");
+
+                // Attempt to get SkillConversationReference using deprecated method.
+                // this catch should be removed once we remove the deprecated method. 
+                // We need to use the deprecated method for backward compatibility.
+#pragma warning disable 618
+                var conversationReference = await _conversationIdFactory.GetConversationReferenceAsync(conversationId, cancellationToken).ConfigureAwait(false);
+#pragma warning restore 618
+                skillConversationReference = new SkillConversationReference
+                {
+                    ConversationReference = conversationReference,
+                    OAuthScope = _getOAuthScope()
```

Comment:
>_getOAuthScope [](start = 33, length = 14)

Is this needed? This needs us to expose OriginatingAudience in BotFrameworkAuthentication class.

Reply:
And we still need to determine this. We really don't want to expose properties. By all means we can extend the functionality of the BotFrameworkAuthentication abstraction. We need to understand exactly why/how this is used and then we can discuss adding more responsibility to the BotFrameworkAuthentication. Exposing properties is the anti-pattern that got us in trouble in the first place.

Reply:
we will need to address this in a separate PR